music21/musicxml/m21ToXml.py

```diff
@@ -3076,6 +3083,77 @@ def mainElementsParse(self):
             # Assumes voices are flat...
             self.parseFlatElements(v, backupAfterwards=backupAfterwards)
 
+    def getVoiceNumberOffset(self) -> int:
+        '''
+            Returns the offset that should be added to any voice numbers in this measure,
+            in order to avoid voice numbers in other simultaneous measures within a staff
+            group.  If this measure is not in a staff group, the offset is 0.
+            This is computed once, and cached, for performance.
+        '''
+        if self.voiceNumberOffset is not None:
+            # return the cached computed offset
+            return self.voiceNumberOffset
+
+        def getMaxVoiceNumber(
+                scoreExp: ScoreExporter,
+                partStaff: stream.Part,
+                measure: stream.Measure) -> int:
+            # find the partExporter for partStaff
+            partStaffExp: t.Optional[PartExporter] = None
+            for partx in scoreExp.partExporterList:
+                if partx.stream is partStaff:
+                    partStaffExp = partx
+                    break
+            if partStaffExp is None:
+                return 0
+
+            maxVoiceNumber: int = partStaffExp.maxVoiceNumbers.get(
+                measure.measureNumberWithSuffix(),
+                0
+            )
+            return maxVoiceNumber
+
+        voiceNumberOffset: int = 0
+
+        partExp: t.Optional[PartExporter] = self.parent
+        if partExp is None:
+            return voiceNumberOffset
+
+        scoreExp: t.Optional[ScoreExporter] = partExp.parent
+        if scoreExp is None:
+            return voiceNumberOffset
+
+        # scoreExp.groupsToJoin is a list of StaffGroups
+        if not scoreExp.groupsToJoin:
+            return voiceNumberOffset
+
+
+        assert(isinstance(self.stream, stream.Measure))
+        myMeasure: stream.Measure = self.stream
+
+        for group in scoreExp.groupsToJoin:
+            if partExp.stream not in group:
+                continue
+
+            for partStaff in group:
+                if partStaff is partExp.stream:
+                    # skip our own PartStaff
+                    continue
+
+                otherMeasure: stream.Measure = partStaff.measure(
+                    myMeasure.measureNumberWithSuffix()
+                )
+                if otherMeasure is None:
+                    # if measures don't have numbers, we can't do this
```

Comment:
I think we can do better than this, since many music21 operations can leave measures unnumbered or poorly numbered, especially if you're doing anything algorithmic, or you're appending streams (see recent list posts). Use your IDE to search for "Measure 0" and you'll find plenty of documented examples of streams all having no measure numbers.

I had a go at an algorithm for matching measure numbers in `processSubsequentPartStaff`. I see two options: shoehorn all of your voice numbering changes into that after-process inside `processSubsequentPartStaff`, which is a little bit like your original approach (sorry) -- OR -- factor out my logic into a generator function kind of like `zip()` that will spit out tuples of measures, eg...
```python
>>> list(zip_measures(p))
[(m, m), (m, None), (m, None), (m, m), ...
```
Let me know if this makes sense. I'm just feeling that if we can't support the use case of no measure numbers, then we're trading the constraint of "hey Greg, make sure your humdrum parser creates voice numbers" for "hey Greg, make sure your humdrum parser creates measure numbers."